### PR TITLE
feat(eslint-plugin): new rule sort-imports-metadata

### PR DESCRIPTION
**Pull request by abaran30**
This pull request introduces the new rule `@angular-eslint/sort-imports-metadata` as described in #1232.

### CODE REVIEW

File: packages/eslint-plugin/src/rules/sort-imports-metadata.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  },
+  defaultOptions: [
+    {
+      locale: DEFAULT_LOCALE,
+    },
+  ],
+  create(context /* [{ locale }] */) {
+    const selectors = [
+      `${Selectors.COMPONENT_CLASS_DECORATOR} Property[key.name!="deps"] > ArrayExpression`,
+      `${Selectors.MODULE_CLASS_DECORATOR} Property[key.name!="deps"] > ArrayExpression`,
```
[the remark this answers is not in the document]
Comment (abaran30) [reply packages/eslint-plugin/src/rules/sort-imports-metadata.ts]: Done! Also made it to check for the `imports` key only.

**Comment by abaran30** [comment]
Update: I added a first version implementation of the new rule and some basic tests for sanity. Perhaps the logic could be more efficient. @rubiesonthesky and @JamesHenry please take a peek when you can and LMK what you think!

P.S. I'll look to add applicable tests from `sort-ngmodule-metadata-arrays` to ensure those are working as expected.

File: packages/eslint-plugin/tests/rules/sort-imports-metadata/cases.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+      ],
+    })
+    export class TestComponent { }
+    `,
+    messageId,
+    annotatedOutput: `
+    @Component({
+      standalone: true,
+      selector: 'test-component',
+      imports: [aModule, bModule, cModule, dModule],
```
Comment (abaran30) [inline packages/eslint-plugin/tests/rules/sort-imports-metadata/cases.ts:141]: How exactly do we want to handle imports listed per line?

**Comment by abaran30** [comment]
Update on how comments are handled: currently, they get stripped away and we're left with the import items only. It is possible to extract the comments and re-insert them where they belong, but it will require a bit of logic to do so.

Will need to spend some time playing around with it.

File: packages/eslint-plugin/tests/rules/sort-imports-metadata/cases.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    })
+    export class TestComponent { }
+    `,
+    messageId,
+    annotatedOutput: `
+    @Component({
+      standalone: true,
+      selector: 'test-component',
+      imports: [
+        // Line comment above aModule
```
Comment (abaran30) [inline packages/eslint-plugin/tests/rules/sort-imports-metadata/cases.ts:162]: I still don't fully understand if/when an `annotatedOutput` also requires the `~~~` annotations found in `annotatedSource`.

File: packages/eslint-plugin/tests/rules/sort-imports-metadata/cases.ts [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+      cModule,
+      dModule
+    ],
+  })
+  export class TestComponent { }
+  `,
+];
+
+export const invalid = [
+  // TODO: This test fails for two reason:
```
Comment (abaran30) [inline packages/eslint-plugin/tests/rules/sort-imports-metadata/cases.ts:47]: This `TODO` starts creeping into formatting territory, which we should limit/avoid.

File: packages/eslint-plugin/tests/rules/sort-imports-metadata/cases.ts [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+  // -         cModule,
+  // -         dModule,
+  // +                 aModule,
+  // +                 bModule,
+  // +                 cModule,
+  // +                 dModule,
+  //
+  // Second error is that the fix receives additional characters at the end of the array:
+  //   dModule,
+  // ],·················
```
Comment (abaran30) [inline packages/eslint-plugin/tests/rules/sort-imports-metadata/cases.ts:60]: Still trying to figure out how these characters come about.

File: packages/eslint-plugin/src/rules/sort-imports-metadata.ts [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+        // Sort the imports
+        importEntriesWithComments.sort((importA, importB) =>
+          importA.importName.localeCompare(importB.importName, locale),
+        );
+
+        const sortedImportsWithCommentsString = `[\n${getSortedImportsWithCommentsAsString(
+          importEntriesWithComments,
+          whitespacePrefix,
+        )}${' '.repeat(nodeForFixer.parent?.loc.start.column as number)}]`;
```
Comment (abaran30) [inline packages/eslint-plugin/src/rules/sort-imports-metadata.ts:144]: This is a bit too magical and in formatting territory...

**Comment by abaran30** [comment]
@JamesHenry updated with the following actions:
- Rebased with latest main
- Pulled some of the auto-fixer logic into a function (passing the `context` instead of `sourceCode`)
- Added applicable tests from existing rule `sort-ngmodule-metadata-arrays`

Ready for review.

P.S. if we still want to test this with a real/large codebase, LMK how we can create a separate release containing this rule to test (I would have to find or create a codebase to work with).

**Comment by abaran30** [comment]
TODO: This rule currently doesn't handle `CallExpression` imports, e.g. `RouterModule.forChild(...)`.

**Comment by abaran30** [comment]
Thank you for sharing @rubiesonthesky! This will work for the check if the imports are sorted. However, adapting the auto-fixer to support this, as it's currently implemented, will get a bit tricky since it essentially builds the new imports list as a string. Will have to play around with it.

**Comment by VictorienTardif** [comment]
Hello, is this PR dropped?
Is there any alternative?
Thx

**Comment by joergbaier** [comment]
It seems a bit dogmatic to exclude rules that improve the consistency and readability of Angular projects. Angular is already a heavily opinionated framework. A rule like this considers the context of Angular code and applies a standard to it. Many new and existing Angular projects will use this package by default, why not provide them with consistency?

These rules fall in the same category:

- `@angular-eslint/template/attributes-order`
- https://github.com/angular-eslint/angular-eslint/pull/1320

Should another tool (prettier plugin) be doing these instead? Applying an order doesn't really "belong" to formatting tools either.

**Comment by abaran30** [comment]
Apologies for the lack of activity here... So we recently discovered that there's at least one case where the order of imports matters for functionality, and that is when simulating a data server using the [In-memory Web API](https://github.com/angular/angular/tree/main/packages/misc/angular-in-memory-web-api). The docs explicitly state, "Always import the `HttpClientInMemoryWebApiModule` *after* the `HttpClientModule` to ensure that the in-memory backend provider supersedes the Angular version." Sorting the imports would break this.

We'll see if there are any other cases like these (feel free to share if you know of any). However, in my opinion as of now, we should not have a rule that would have to dance around certain exceptions like these in order to work for most/all users.

**Comment by abaran30** [comment]
Closing this PR, will add a comment to the issue (#1232).

**Pull request closed without merging**